From: SeongJae Park <sj@kernel.org>
Date: Sat, 14 Feb 2026 12:53:33 -0800
Subject: [PATCH] mm/damon/core: allow only power of two min_region_sz

The value is used by DAMON core layer for aligning DAMON region.  The
align functions supporty only power of two alignments.  But DAMON core
API callers can set min_region_sz to an arbitrary number.  Users can set
it indirectly, using addr_unit.

Even if the alignment is not properly set, it doesn't cause crash-like
issues.  But it makes monitoring and DAMOS behavior difficult to expect
and understand, effectiely broken.

Fix this by disallowing min_region_sz that is not a power of two.
Add the check to damon_commit_ctx(), as all DAMON API callers who set
min_region_sz uses the function.

Fixes: d8f867fa0825 ("mm/damon: add damon_ctx->min_sz_region")
Cc: <stable@vger.kernel.org> # 6.18.x
Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 9f4e2ac8ac1d2..cf3ad89c12287 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -1495,6 +1495,9 @@ int damon_commit_ctx(struct damon_ctx *dst, struct damon_ctx *src)
 {
 	int err;
 
+	if (!is_power_of_2(src->min_region_sz))
+		return -EINVAL;
+
 	err = damon_commit_schemes(dst, src);
 	if (err)
 		return err;
-- 
2.47.3

